### ISSUE TITLE
fix: inconsistent torch.enable_grad() usage across optimizers

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#70]
## Problem

The optimizers have inconsistent contracts regarding autograd context:

- **LevenbergMarquardt, ExtendedKalmanFilter, KalmanFilter**: wrap `closure = torch.enable_grad()(closure)` inside `step()`.
- **Newton, Annealing, Metropolis, Genetic**: do not wrap the closure at all.

This means a user who guards their training loop with `torch.no_grad()` (a common pattern for evaluation + optimization) gets different behavior per optimizer:
- LM/EKF/KF work correctly (the closure re-enables grad internally).
- Newton/Annealing/Metropolis/Genetic silently produce zero gradients and incorrect updates.

## Proposal

Standardize all optimizers to use the same pattern:

```python
@torch.no_grad()
def step(self, closure):
    with torch.enable_grad():
        # evaluate closure and compute gradients here
        ...
    # weight updates happen in no_grad context
    ...
```

This is the pattern used by `torch.optim.SGD`, `torch.optim.Adam`, etc. It ensures:
1. Weight updates don't build a graph (no memory leak).
2. The closure always runs with grad enabled, regardless of the caller's context.
3. The contract is the same for every optimizer.

**Comment by kovacoj** [comment]
## Proposed fix

Standardize all optimizers to use `@torch.no_grad()` on `step()` with `torch.enable_grad()` only for the closure and gradient computation:

**Pattern:**
```python
@torch.no_grad()
def step(self, closure):
    assert len(self.param_groups) == 1
    # Closure and explicit grad calls re-enable grad locally
    with torch.enable_grad():
        loss = closure()
        grads = grad(loss, params, ...)
    # Weight updates happen in no_grad context — no graph retained
    self.update_weights(direction)
```

**Per-optimizer changes:**

| Optimizer | Current | Proposed |
|---|---|---|
| Newton | No `no_grad`, no `enable_grad` | `@torch.no_grad()` + `with torch.enable_grad():` around closure/grad |
| LM | `torch.enable_grad()(closure)` | `@torch.no_grad()` + keep `torch.enable_grad()` wrapper |
| EKF | `torch.enable_grad()(closure)` | `@torch.no_grad()` + keep `torch.enable_grad()` wrapper |
| KF | `torch.enable_grad()(closure)` | `@torch.no_grad()` + keep `torch.enable_grad()` wrapper |
| Annealing | No `no_grad`, no `enable_grad` | `@torch.no_grad()` + `with torch.enable_grad():` around closure |
| Metropolis | No `no_grad`, no `enable_grad` | `@torch.no_grad()` + `with torch.enable_grad():` around closure |
| Genetic | No `no_grad`, no `enable_grad` | `@torch.no_grad()` + `with torch.enable_grad():` around `directional()` |

**This supersedes #66** — the `@torch.no_grad()` decorator is part of this standardization, and the `torch.enable_grad()` wrapper is the other half.

**Key behavioral change:** If a user calls `optimizer.step(closure)` inside a `torch.no_grad()` block, the closure will still execute with grad enabled. This matches PyTorch's standard optimizer contract (SGD, Adam, etc.) and is the expected behavior.
